### PR TITLE
Saner and consistent YAML fields in config

### PR DESCRIPTION
**Pull request by gouthamve**
Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ingester/ingester.go [hunk 1]
```diff
@@ -48,17 +48,17 @@ type Config struct {
...
-	MaxChunkAge       time.Duration
-	ChunkAgeJitter    time.Duration
-	ConcurrentFlushes int
-	SpreadFlushes     bool
-
-	RateUpdatePeriod time.Duration
+	FlushCheckPeriod  time.Duration `yaml:"flush_period"`
+	RetainPeriod      time.Duration `yaml:"retain_period"`
+	MaxChunkIdle      time.Duration `yaml:"max_chunk_idle"`
+	MaxStaleChunkIdle time.Duration `yaml:"max_stale_chunk_idle"`
```
Comment (pstibrany) [inline pkg/ingester/ingester.go]:
```suggestion
	MaxChunkIdle      time.Duration `yaml:"max_chunk_idle_time"`
	MaxStaleChunkIdle time.Duration `yaml:"max_stale_chunk_idle_time"`
```

File: pkg/ingester/ingester.go [hunk 2]
```diff
@@ -803,9 +803,9 @@ func (i *Ingester) Watch(in *grpc_health_v1.HealthCheckRequest, stream grpc_heal
 	return status.Error(codes.Unimplemented, "Watching is not supported")
 }
 
-// ReadinessHandler is used to indicate to k8s when the ingesters are ready for
-// the addition removal of another ingester. Returns 204 when the ingester is
-// ready, 500 otherwise.
+// CheckReady is the readiness handler is used to indicate to k8s when the ingesters
```
Comment (pstibrany) [inline pkg/ingester/ingester.go]:
```suggestion
// CheckReady is the readiness handler used to indicate to k8s when the ingesters
```

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -803,9 +803,9 @@ func (i *Ingester) Watch(in *grpc_health_v1.HealthCheckRequest, stream grpc_heal
...
+// are ready for the addition removal of another ingester. Returns 204 when the
+// ingester is ready, 500 otherwise.
```
Comment (pstibrany) [inline pkg/ingester/ingester.go]:
```suggestion
// are ready for the addition removal of another ingester.
```

HTTP status codes are not returned from here anymore. (Fixing my code change via your PR :))

File: pkg/querier/frontend/worker.go [hunk 4]
```diff
@@ -31,9 +31,9 @@ var (
 
 // WorkerConfig is config for a worker.
 type WorkerConfig struct {
-	Address           string
-	Parallelism       int
-	DNSLookupDuration time.Duration
+	Address           string        `yaml:"address"`
```
Comment (pstibrany) [inline pkg/querier/frontend/worker.go]: What is address? URL? Can be `frontend_url` instead

File: pkg/querier/querier.go [hunk 5]
```diff
@@ -21,12 +21,12 @@ import (
...
 type Config struct {
-	MaxConcurrent        int
-	Timeout              time.Duration
-	Iterators            bool
-	BatchIterators       bool
-	IngesterStreaming    bool
-	MaxSamples           int
+	MaxConcurrent        int           `yaml:"max_concurrent"`
+	Timeout              time.Duration `yaml:"timeout"`
+	Iterators            bool          `yaml:"iterators"`
```
Comment (pstibrany) [inline pkg/querier/querier.go]: This sounds like "number of iterators" to me, not as a boolean flag.

File: pkg/ring/kv/consul/client.go [hunk 6]
```diff
@@ -35,12 +35,12 @@ var (
 
 // Config to create a ConsulClient
 type Config struct {
-	Host              string
-	ACLToken          string
-	HTTPClientTimeout time.Duration
-	ConsistentReads   bool
-	WatchKeyRateLimit float64 // Zero disables rate limit
-	WatchKeyBurstSize int     // Burst when doing rate-limit, defaults to 1
+	Host              string        `yaml:"host"`
```
Comment (pstibrany) [inline pkg/ring/kv/consul/client.go]: `url`? I think it would be nice to be consistent to use `url` in the names for fields which are actually URLs.

Comment (jtlisi) [reply pkg/ring/kv/consul/client.go]: I agree, I think this should be consistent where relevant

File: pkg/ruler/ruler.go [hunk 7]
```diff
@@ -52,23 +52,35 @@ var (
...
-	Ring             RingConfig
-	FlushCheckPeriod time.Duration
+	// This is used for template expansion in alerts; must be a valid URL.
+	ExternalURL flagext.URLValue `yaml:"external_url"`
+	// How frequently to evaluate rules by default.
+	EvaluationInterval time.Duration `yaml:"evaluation_interval"`
+	// How frequently to poll for updated rules.
+	PollInterval time.Duration `yaml:"poll_interval"`
+	// Rule Storage and Polling configuration.
+	StoreConfig RuleStoreConfig `yaml:"store_config"`
```
Comment (jtlisi) [inline pkg/ruler/ruler.go]: Can you call this yaml field `storage` to match the flag prefix it adds?

File: pkg/ingester/ingester.go (hunk 1 above)
Comment (jtlisi) [reply pkg/ingester/ingester.go]: @pstibrany would you suggest we suffix all duration values like this with `_time`. Similar to how counters in Prometheus are suffixed with `_total`. I think that might be a little clunky but does provide value.

File: pkg/alertmanager/multitenant.go [hunk 8]
```diff
@@ -96,20 +96,20 @@ func init() {
...
+	Peers                flagext.StringSlice `yaml:"peers"`
+	PeerTimeout          time.Duration       `yaml:"peer_timeout"`
 
-	FallbackConfigFile string
-	AutoWebhookRoot    string
+	FallbackConfigFile string `yaml:"fallback_config_file"`
+	AutoWebhookRoot    string `yaml:"auto_webhook_root"`
 
-	Store AlertStoreConfig
+	Store AlertStoreConfig `yaml:"store"`
```
Comment (jtlisi) [inline pkg/alertmanager/multitenant.go]: Can this also be a `storage` field? Similar to the ruler?

File: pkg/ring/kv/consul/client.go (hunk 6 above)
Comment (pstibrany) [reply pkg/ring/kv/consul/client.go]: Actually, this is not a URL :-( Expected format here is host:port.

File: pkg/querier/frontend/worker.go (hunk 4 above)
Comment (pstibrany) [reply pkg/querier/frontend/worker.go]: We should specify expected format of address / host fields, especially if it is not just hostname.

File: pkg/chunk/aws/dynamodb_storage_client.go [hunk 9]
```diff
@@ -97,13 +97,13 @@ func init() {
...
-	ApplicationAutoScaling flagext.URLValue
-	Metrics                MetricsAutoScalingConfig
-	ChunkGangSize          int
-	ChunkGetMaxParallelism int
+	DynamoDB               flagext.URLValue         `yaml:"dynamodb_url"`
+	APILimit               float64                  `yaml:"api_limit"`
+	ThrottleLimit          float64                  `yaml:"throttle_limit"`
+	ApplicationAutoScaling flagext.URLValue         `yaml:"application_autoscaling_url"`
+	Metrics                MetricsAutoScalingConfig `yaml:"metrics"`
+	ChunkGangSize          int                      `yaml:"chunk_gang_size"`
```
Comment (pracucci) [inline pkg/chunk/aws/dynamodb_storage_client.go]: I believe its CLI flag name is not correct. It should end with `.chunk-gang-size`

File: pkg/chunk/aws/dynamodb_storage_client.go [hunk 10]
```diff
@@ -97,13 +97,13 @@ func init() {
...
+	ChunkGetMaxParallelism int                      `yaml:"chunk_get_max_parallelism"`
```
Comment (pracucci) [inline pkg/chunk/aws/dynamodb_storage_client.go]: I believe its CLI flag name is not correct. It should end with `.chunk-get-max-parallelism`

File: pkg/ring/kv/consul/client.go [hunk 11]
```diff
@@ -35,12 +35,12 @@ var (
...
+	ACLToken          string        `yaml:"acl_token"`
```
Comment (pracucci) [inline pkg/ring/kv/consul/client.go]: The CLI flag should end with `.acl-token`. Then remember to update `arguments.md`

**Comment by bboreham** [comment]
>  booleans names like enable_ (not disable)

Historically these were chosen so the zero value (`false`) is the default.
At some point we went to mandatory initialization of config structs, so maybe that reason has gone, but I don't think there is a strong argument just on consistency.

For many flags the default is "what Cortex did earlier" but should be changed to "what most people would use".  I think this type of change is much more important.

File: docs/configuration/config-file-reference.md [hunk 12]
```diff
@@ -1345,24 +1346,24 @@ The `storage_config` configures where Cortex stores the data (chunks storage eng
 [engine: <string> | default = "chunks"]
 
 aws:
-  dynamodbconfig:
+  dynamodb_config:
```
Comment (pracucci) [inline docs/configuration/config-file-reference.md]: Why not simply `dynamodb`?

**Pull request merged**